### PR TITLE
Implementing exponentiation by squaring for matrix power

### PR DESCRIPTION
**Pull request by abdalazizrashid**
Issue #115 raised the attention that the current implementation of matrix power is implemented naively; thus, here is an implementation that uses exponentiation by squaring, which is significantly faster.

### CODE REVIEW

**Comment by ferrine** [comment]
I think a better way is to implement this via [http://deeplearning.net/software/theano/extending/extending_theano.html](theano.Op)

**Review by brandonwillard — COMMENTED** [review]
@abdalazizrashid, this looks significantly better than the current implementation!

@ferrine is correct, because&mdash;for instance&mdash;the current implementation won't allow us to use symbolic values for the `n` argument (e.g. `matrix_power(M, tt.iscalar())`).  There are two basic ways around this: 
1. as @ferrine suggests, we create a `MatrixPower` `Op`, or
2. we reproduce the steps in your current implementation using Theano (e.g. use `IfElse` or `tt.switch` `Op`s for the conditional statements and `scan` for the `while`-loop).

Approach 1. can be fairly copy-paste until we get to the `Op.grad` and `Op.c_code` implementations.  The former seems fairly easy to implement in this case, but the latter can be annoying.  We don't have to implement `Op.c_code`, though.  (Same goes for `Op.grad`, but that one really _should_ be implemented.)

Approach 2. will basically give us everything that a fully implemented `Op` would (e.g. the gradient and C code will be available, as long as the Theano `Op`s we use have them implemented), but there could be a performance hit due to the use of `scan`.

Now that I think about it a little more, _I think Approach 2. is the better option_.  If we take this approach and then invest our efforts into debugging and fixing any performance problems that it may have, those efforts will pay off much more, since they'll apply to anything else that uses the same underlying `Op`s.  In other words, *if we improve `scan` just a little bit, we may end up improving Theano a lot*.

In the meantime, we can go forward with this implementation&mdash;after the tests succeed&mdash;and create another issue for further improvements, since this is already a big improvement over the current approach&mdash;and it doesn't introduce any limitations that weren't already there.  If you want to try one of the other approaches, feel free to rebase this branch with those changes.


Also, you can automatically avoid most of these code style errors by installing [`pre-commit`](https://pre-commit.com/#install) and running `pre-commit install` in this project's root directory.  That will make sure that the checks are run at every commit.

File: theano/tensor/nlinalg.py [hunk 1]
```diff
@@ -663,17 +663,43 @@ def perform(self, node, inputs, outputs):
 
 
 def matrix_power(M, n):
-    """
+    r"""
     Raise a square matrix to the (integer) power n.
+    This implementation uses exponentiation by squaring which is
+    significantly faster than the naive implementation.
+    The time complexity for exponentiation by squaring is
+    :math: `\mathcal{O}((n \log M)^k`
```
Comment (brandonwillard) [inline theano/tensor/nlinalg.py:671]: Looks like this is missing a closing parenthesis.

**Review by brandonwillard — CHANGES_REQUESTED** [review]
The [test needs a few more cases](https://coveralls.io/builds/34288521/source?filename=theano%2Ftensor%2Fnlinalg.py#L684).

Note: you don't need to compile and compute the values; `theano.dot` is a well tested `Op`, and compiling graphs and C-code using `function` is overkill.  It's a very big problem with this old test suite and has led to unnecessarily long test durations.

Since we need to confirm that the underlying algorithm is correct, we can't simply check that the resulting symbolic graph matches our expectations, but we can at least avoid the compilation parts by using the test values.  Just set `A.tag.test_value` to the NumPy test values and get the result from `q.get_test_value()`.

File: tests/tensor/test_nlinalg.py [hunk 2]
```diff
@@ -555,14 +557,12 @@ def test_non_square_matrix(self):
 
     def test_negative_matrix(self):
```
Comment (brandonwillard) [inline tests/tensor/test_nlinalg.py:558]: Couldn't this test be covered by simply adding a `-1` to the test parameters in `test_numpy_compare` above?

File: tests/tensor/test_nlinalg.py [hunk 3]
```diff
@@ -533,16 +536,15 @@ def test_wrong_rcond_dimension(self):
...
 
 class TestMatrix_power:
-    def test_numpy_compare(self):
+    @pytest.mark.parametrize("n", [0, 1, 2, 3, 4, 5, 11])
+    def test_numpy_compare(self, n):
         rng = np.random.RandomState(utt.fetch_seed())
-        A = tensor.matrix("A", dtype=theano.config.floatX)
-        Q = matrix_power(A, 3)
-        fn = function([A], [Q])
         a = rng.rand(4, 4).astype(theano.config.floatX)
```
Comment (brandonwillard) [inline tests/tensor/test_nlinalg.py:542]: There's really no need to use random numbers in these examples.  We're better off using a single representative 2 x 2 matrix for these tests (e.g. one that's neither idempotent nor vanishes too quickly).

**Review by brandonwillard — CHANGES_REQUESTED** [review]

**Comment by brandonwillard** [comment]
It looks like [some tests are failing](https://github.com/pymc-devs/Theano-PyMC/pull/117/checks?check_run_id=1283549857#step:6:1734) because of that global test value setting.  There are some brittle tests that check the lengths of captured warning messages and those messages now include the new test value warnings.

File: tests/tensor/test_nlinalg.py (hunk 2 above)
Comment (abdalazizrashid) [reply tests/tensor/test_nlinalg.py:558]: Hmm, yeah actually make sense. Thanks!

File: tests/tensor/test_nlinalg.py (hunk 3 above)
Comment (abdalazizrashid) [reply tests/tensor/test_nlinalg.py:542]: Well, the way I approached this is basically by making minimal changes to the codebase.

**Comment by abdalazizrashid** [comment]
I think something is wrong; the coverage test took forever!

File: tests/tensor/test_nlinalg.py (hunk 3 above)
Comment (brandonwillard) [reply tests/tensor/test_nlinalg.py]: My comment was actually about big-picture changes that I would like to start making in Theano.  Your approach is definitely the correct one, so no problems there.

**Comment by brandonwillard** [comment]
> I think something is wrong; the coverage test took forever!

That's due to a bug introduced by #119.  I just reverted that change, so, if you rebase and push, a coverage report should be available.

**Review by brandonwillard — CHANGES_REQUESTED** [review]

**Review by brandonwillard — APPROVED** [review]

**Pull request merged**